### PR TITLE
Adding glomap support to ns-process-data, requires latest version of COLMAP

### PR DESCRIPTION
**Pull request by AntonioMacaronio**
**Problems and Background**
- GLOMAP is orders of magnitudes faster than COLMAP On average, it finds camera poses and intrinsics 100x faster than COLMAP, and it would great to support this in Nerfstudio and already has been [requested](https://github.com/nerfstudio-project/nerfstudio/issues/3402)

**Overview of Changes**
- Adding a GLOMAP toggle in `ns-process-data` script, which uses the GLOMAP mapper

### CODE REVIEW

**Comment by kerrj** [comment]
This is great! what is the difference between the `sfm_tool` option and `glomap_toggle`? It would also be good to write documentation or link to installation instructions and the like

File: nerfstudio/process_data/colmap_utils.py [hunk 1]
```diff
@@ -153,12 +154,20 @@ def run_colmap(
...
     sparse_dir = colmap_dir / "sparse"
     sparse_dir.mkdir(parents=True, exist_ok=True)
-    mapper_cmd = [
-        f"{colmap_cmd} mapper",
-        f"--database_path {colmap_dir / 'database.db'}",
-        f"--image_path {image_dir}",
-        f"--output_path {sparse_dir}",
-    ]
+    if glomap_toggle:
+        mapper_cmd = [
```
Comment (pablovela5620) [inline nerfstudio/process_data/colmap_utils.py:158]: Seems like this could be more consice if you did something like
```python
 mapper_cmd = [
    f"{'glomap' if glomap_toggle else colmap_cmd} mapper",
    f"--database_path {colmap_dir / 'database.db'}",
    f"--image_path {image_dir}",
    f"--output_path {sparse_dir}",
]
```

**Comment by hyqshr** [comment]
Any update on this?

**Comment by jb-ye** [comment]
@AntonioMacaronio Do you want to rebase your changes on top main? I think this one is close to merge.

**Comment by AntonioMacaronio** [comment]
@jb-ye Sure! I'll need to do one last round of testing and I'll fix the build errors and this should be good to go

**Comment by pablovela5620** [comment]
> @jb-ye update on this: there is currently a bug with how the cameras are placed in the viewer and during training. In the colmap visualizer, the camera poses are fine and look accurate. However, during training in the nerfstudio viewer all the cameras get placed in 1 spot at the origin. I'll have to look into the nerfstudio_dataparser.py to see the issue with this
@AntonioMacaronio 
I think (?) this is may be due to lack of normalization. What commit of glomap are you using? They recently add this fix in this commit
https://github.com/colmap/glomap/commit/d41b747407b6c5ce564e3936d0f2f2ea52f89dae
I had a similar problem but it got fixed this this update
https://github.com/colmap/glomap/issues/128#issuecomment-2474941872

**Comment by AntonioMacaronio** [comment]
@pablovela5620 I just updated my glomap to master and tried this, but the cameras still get placed at the origin; have you had any success with this PR? Sometimes it doesn't get placed at the origin and the poses look accurate but because it's finnicky so I'm hesitant on merging

**Comment by pablovela5620** [comment]
It turns out there's another issue, and you're right to be hesitant! I've been doing some debugging, and as you stated, even with the latest PR, you get the weird clustering effect. I basically ran colmap/glomap (not with this pr but with my own implementation using a different feature+matcher combo). I found that glomap can have some catastrophic failures for a small number of cameras out of the whole group. I linked a video showing this, where I'm comparing a reconstruction of 250 images using the aliked feature detector and light glue matcher with colmap and then with glomap. 

Basically one of the cameras is WAY out of bounds leading to the clustering at the origin, one or two cameras end up being 10x farther out due to a failure in point triangulation(?). I'm going to open an issue on the glomap repo

[glomap-failure-featureless.webm](https://github.com/user-attachments/assets/cf9b78de-3563-4977-ad82-f5b8fb3d1261)

**Comment by AntonioMacaronio** [comment]
> Basically one of the cameras is WAY out of bounds leading to the clustering at the origin, one or two cameras end up being 10x farther out due to a failure in point triangulation(?).
I saw your comment with the video of the COLMAP gui and I see the exact same thing with my COLMAP gui. Most of the cameras are close together and accurate but one of them is very far away. What's puzzling is that Nerfstudio compacts all the close cameras to the origin. Even if one of the cameras is far away, because Nerfstudio is scaling the distances between the cameras, it should still remain proportional (in my thinking), so I'm still not sure why that happens

**Comment by pablovela5620** [comment]
Yea, I'm not sure why it fails in Nerfstudios case. Might be worth looking into the normalization code that Glomap added to see how they fixed it. In either case, I think until these outlier cases get resolved its probably not good to merge this is. The massive outliers completely ruin the reconstruction into something nonsensical. Maybe there could be some heuristic value that tries to deal with outliers (like projection error or if the camera is 10x further from the center of mass or something) but those still feel very hacky

**Comment by Jourdelune** [comment]
Any news?
